### DISCUSSION TITLE
Arduino with Firmata and Bonsai AnalogRead too slow

### DISCUSSION THREAD

**Discussion by maimechevee** [bonsai-rx/bonsai#2061] [Q&A]
Hello, 
I am trying to read a 20Hz pulse off an analog pin of an Arduino Uno. The Arduino is loaded with StandardFirmata and Bonsai sketch is attached. Any suggestion why this set up wouldn't work? Thank you!
Maxime
[isxSync_analog2024-11-21T14_24_57.csv](https://github.com/user-attachments/files/17871922/isxSync_analog2024-11-21T14_24_57.csv)
[isxSync_Timestamps_2024-11-21T14_24_57.csv](https://github.com/user-attachments/files/17871924/isxSync_Timestamps_2024-11-21T14_24_57.csv)
![Uploading IMG_7472.jpg…]()

**Comment by bruno-f-cruz** [comment]
I don't think your attachments were uploaded successfully.

**Comment by maimechevee** [comment]
Sorry about that..
![IMG_7472](https://github.com/user-attachments/assets/7c5b99c1-2b09-4291-9005-2a071efbb2f6)

**Reply by bruno-f-cruz** [reply]
I am probably missing something. It appears like your workflow should work to acquire analog input data. What is the problem exactly? Is it just an issue to increase the adc sampling rate? In that case, you can reduce the value of `samplingInterval` in the firmata sketch.

```c++
/* timer variables */
unsigned long currentMillis;        // store the current value from millis()
unsigned long previousMillis;       // for comparison with currentMillis
unsigned int samplingInterval = 19; // how often to run the main loop (in ms)
```

**Reply by maimechevee** [reply]
Hi Bruno, 
Thanks for your suggestions - I did indeed have the samplingInterval = 19. However, reducing the value did not change the behavior I described above. 

For now, I have switched to using SimpleDigitalFirmata instead. This gives me a precision of about 5ms (timestamps look like 48,49,47,53, 49, 48, ... milliseconds). That's what I was using at first, but in an effort to get exactly 50ms I tried to read the pulses off an analog pin and switched to StandardFirmata... without success so far. 

So this works enough for my purpose - but I'd love to understand why I can't get Bonsai to precisely record 20Hz pulses :) 

Thanks!

Maxime

**Reply by bruno-f-cruz** [reply]
> This gives me a precision of about 5ms (timestamps look like 48,49,47,53, 49, 48, ... milliseconds). That's what I was using at first, but in an effort to get exactly 50ms I tried to read the pulses off an analog pin and switched to StandardFirmata... without success so far.

If you are trying to rely on the OS clock to timestamp your data, you will not easily be able (if at all) get the precision you are looking for (i.e. "exactly 50ms"). If you want strict timestamping you should timestamp in hardware instead. This is not a limitation of Bonsai but instead due to the [non-RTOS nature of common computer operating systems](https://www.tutorialspoint.com/difference-between-rtos-and-os). 

I am still a bit confused as to what "behavior" you are refering to exactly. Reducing the sampling rate (within the allowed regime) should produce more samples per second.

>However, reducing the value did not change the behavior I described above.

**Reply by maimechevee** [reply]
Hi Bruno, 

I investigated a little more following the link you sent, and I understand that this needs to be addressed on the hardware itself... For now, this works well enough for my application so I'm going to keep using SimpleDigitalFirmata until I have to revisit the issue. 

Thanks again!
